packages/flutter/lib/src/widgets/fade_in_image.dart

```diff
@@ -0,0 +1,273 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/services.dart';
+
+import 'basic.dart';
+import 'framework.dart';
+import 'image.dart';
+import 'ticker_provider.dart';
+
+/// An image that shows a [placeholder] image while the target [image] is
+/// loading, then fades in the new image when it loads.
+///
+/// Use this class to display long-loading images, such as [new Image.network],
+/// so that the image appears on screen with a graceful animation rather than a
+/// sudden jerk.
+///
+/// This image ignores the [placeholder]'s and [image]'s `color` and `blendMode`
+/// properties in order to efficiently perform the cross-fade animation.
+///
+/// If the [image]'s [ImageProvider] returns [ImageInfo] synchronously, such as
+/// when the image has been loaded and cached, the [image] is displayed
+/// immediately and the [placeholder] is never displayed.
+///
+/// [fadeOutDuration] and [fadeOutCurve] control the fade-out animation of the
+/// placeholder.
+///
+/// [fadeInDuration] and [fadeInCurve] control the fade-in animation of the
+/// target [image].
+///
+/// Example:
+///
+/// ```dart
+/// return new FadeInImage(
+///   placeholder: new Image.memory(bytes),
+///   image: new Image.network('https://yourbackend.com/image.png'),
+/// );
+/// ```
+///
+/// Prefer [placeholder] that's already cached so that it can be displayed in
+/// the frame in which the [FadeInImage] is built.
+class FadeInImage extends StatefulWidget {
+  const FadeInImage({
+    Key key,
+    @required this.placeholder,
+    @required this.image,
+    this.fadeOutDuration: const Duration(milliseconds: 300),
+    this.fadeOutCurve: Curves.easeOut,
+    this.fadeInDuration: const Duration(milliseconds: 700),
+    this.fadeInCurve: Curves.easeIn,
+  }) : super(key: key);
+
+  /// Image displayed while the target [image] is loading.
+  final Image placeholder;
+
+  /// The target image that is displayed.
+  final Image image;
```

Comment:
We shouldn't use widgets here, because we don't really use them as widgets. (For example, the key gets ignored.)
We should just use image providers, then provide some default constructors that pair up nicely, for example memory+asset, memory+network, asset+network.